### DISCUSSION TITLE
export as .npy file

### DISCUSSION THREAD

**Discussion by lingoodluck** [simnibs/simnibs#453] [9. Results and analysis]
I am a beginner who has just started using SimNibs, and I have two questions that I don't quite understand:

I want to generate a .npy file that represents the electric field magnitude in a specified shape. Is there any way to do this?
How can I export a conductivity map or segmented label map that includes at least the scalp, compact bone, spongy bone, gm，and white matter, also represented as a .npy file?

**Comment by TorgeW** [comment]
A few questions here:
1. Do you want the E-field magnitude in the mesh tetrahedra or in voxels?
2. When you are talking about a conductivity map / segmented label map what do you mean exactly? Just the conductivity / label per tetrahedra of the mesh or per voxel?

**Reply by lingoodluck** [reply]
They are all based on voxels and are all three-dimensional. The slice images are similar to this:
![图片1](https://github.com/user-attachments/assets/086a886b-8d39-4441-b6a8-872e91dbbbbb)
![图片2](https://github.com/user-attachments/assets/cdb2587c-e69e-410b-bc4d-bc5a0e46e265)

**Reply by TorgeW** [reply]
https://simnibs.github.io/simnibs/build/html/documentation/sim_struct/session.html#session-doc

When running the simulation, set map_to_vol to True to get the fields inside the volume as a nifti file. You can then load that file with for example https://nipy.org/nibabel/ and save it as a .npy file. 

For the tissues, you can use the final_tissues.nii.gz file in your charm output folder. It contains the tissues that are mapped from the mesh into voxel space.

**Reply by lingoodluck** [reply]
Thank you very much for your help!

**Comment by lingoodluck** [comment]
If I use this piece of code, what should I do?I want to know if I can directly save TImax, and whether the final_tissues.nii.gz file contains conductivity or tissue labels?
```
S = sim_struct.SESSION()
S.subpath = 'D:\Program Files\SimNIBS4.1.0\simnibs4_examples\m2m_ernie'
S.pathfem = 'E:\simnibs_test\TI_test2'  # Directory for the simulation

# specify first electrode pair
tdcs = S.add_tdcslist()
tdcs.currents = [0.001, -0.001]  # Current flow though each channel (A)

electrode = tdcs.add_electrode()
electrode.channelnr = 1
electrode.centre = 'F5'
electrode.shape = 'rect'
electrode.dimensions = [12,12]  # diameter in [mm]
electrode.thickness = 1#[2,1,2]  # 1 mm thickness
#electrode.dimensions_sponge = [20,20]


electrode = tdcs.add_electrode()
electrode.channelnr = 2
electrode.centre = 'P5'
electrode.shape = 'rect'
electrode.dimensions = [12,12]  # diameter in [mm]
electrode.thickness = 1#[2,1,2]  # 1 mm thickness
#electrode.dimensions_sponge = [20,20]

# specify second electrode pair
tdcs = S.add_tdcslist(deepcopy(tdcs))
tdcs.currents = [0.001, -0.001]
tdcs.electrode[0].centre = 'F6'
tdcs.electrode[1].centre = 'P6'


run_simnibs(S)

"""
    generate the TI field from the simulation results
"""
m1 = mesh_io.read_msh(os.path.join(S.pathfem, 'ernie_TDCS_1_scalar.msh'))
m2 = mesh_io.read_msh(os.path.join(S.pathfem, 'ernie_TDCS_2_scalar.msh'))

# remove all tetrahedra and triangles belonging to the electrodes so that
# the two meshes have same number of elements
#删除所有属于电极的四面体和三角形，使两个网格的元素数量相同
tags_keep = np.hstack((np.arange(ElementTags.TH_START, ElementTags.SALINE_START - 1),
                       np.arange(ElementTags.TH_SURFACE_START, ElementTags.SALINE_TH_SURFACE_START - 1)))
m1 = m1.crop_mesh(tags=tags_keep)
m2 = m2.crop_mesh(tags=tags_keep)

# calculate the maximal amplitude of the TI envelope
ef1 = m1.field['E']
ef2 = m2.field['E']
print("Shape of ef1:", ef1.value.shape)
print("Shape of ef2:", ef2.value.shape)
TImax = TI.get_maxTI(ef1.value, ef2.value)

print("TImax",TImax.shape)
# 将 TImax 保存为 .npy 文件
np.save(os.path.join(S.pathfem, 'TI.npy'), TImax)
# make a new mesh for visualization of the field strengths
# and the amplitude of the TI envelope
#创建新的网格，以便可视化场强和 TI 包络线的振幅
mout = deepcopy(m1)
mout.elmdata = []
mout.add_element_field(ef1.norm(), 'magnE - pair 1')
mout.add_element_field(ef2.norm(), 'magnE - pair 2')
mout.add_element_field(TImax, 'TImax')
mesh_io.write_msh(mout, os.path.join(S.pathfem, 'TI.msh'))
v = mout.view(
    visible_tags=[1001, 1002,1003,],
    visible_fields='TImax',
)
v.write_opt(os.path.join(S.pathfem, 'TI.msh'))
mesh_io.open_in_gmsh(os.path.join(S.pathfem, 'TI.msh'), True)

```

**Reply by TorgeW** [reply]
You would add this before run_simnibs(S)

```
S.map_to_vol = True
S.tissues_in_niftis = 'all'
```

final_tissues.nii.gz contains tissue labels

**Reply by lingoodluck** [reply]
TI is calculated after simulation. Can S.map_to_vol = True also save the value of TImax?

**Reply by TorgeW** [reply]
https://github.com/simnibs/simnibs/blob/master/simnibs/utils/transformations.py#L559

You should be able to use this function to interpolate the values from the mesh to voxel space.

**Reply by lingoodluck** [reply]
Thank you for your patient answers！ I will give it a try.
